### PR TITLE
bloom: use Span instead of std::vector for `insert` and `contains`

### PR DESCRIPTION
**Pull request by fanquake**
This is #18985 rebased, with the most recent comments addressed.

> We can avoid many unnecessary std::vector allocations by changing
CBloomFilter to take Spans instead of std::vector's for the `insert`
and `contains` operations.

> CBloomFilter currently converts types such as CDataStream and uint256
to std::vector on `insert` and `contains`. This is unnecessary because
CDataStreams and uint256 are already std::vectors internally. We just
need a way to point to the right data within those types. Span gives
us this ability.

### CODE REVIEW

File: src/bloom.h [hunk 1]
```diff
@@ -66,11 +66,11 @@ class CBloomFilter
 
     SERIALIZE_METHODS(CBloomFilter, obj) { READWRITE(obj.vData, obj.nHashFuncs, obj.nTweak, obj.nFlags); }
 
-    void insert(const std::vector<unsigned char>& vKey);
+    void insert(Span<const unsigned char> vKey);
     void insert(const COutPoint& outpoint);
     void insert(const uint256& hash);
```
[the remark this answers is not in the document]
Comment (MarcoFalke) [reply src/bloom.h:71]: Indeed: https://github.com/bitcoin/bitcoin/pull/18985#discussion_r683582348

**Comment by jonatack** [comment]
Concept ACK

**Comment by laanwj** [comment]
Thanks for picking this up. Concept ACK.

File: src/bloom.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (fanquake) [reply src/bloom.h]: Thought I'd done this in my branch already. Fixed.

**Comment by fanquake** [comment]
Fixed up the change that I should have already made, and added minor bloom cleanup commits.

**Comment by sipa** [comment]
Code review ACK a11da7541148b5bb8e293c0ee49b2856a6628099

**Comment by laanwj** [comment]
Code review ACK a11da7541148b5bb8e293c0ee49b2856a6628099

**Pull request closed without merging**